packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,346 @@
+          node,
+        ) as NodeWithTypeParameters;
+
+        const checker = parserServices.program.getTypeChecker();
+        const counts = countTypeParameterUsage(checker, tsNode);
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const identifierCounts = counts.get(typeParameter.name);
+          if (
+            !identifierCounts ||
+            identifierCounts.general > 2 ||
+            identifierCounts.return > 2 ||
+            (identifierCounts.general === 2 && identifierCounts.return) ||
+            (identifierCounts.return === 2 && identifierCounts.general)
+          ) {
+            continue;
+          }
+
+          context.report({
+            data: {
+              name: typeParameter.name.text,
+            },
+            node: parserServices.tsNodeToESTreeNodeMap.get(typeParameter),
+            messageId: 'sole',
+          });
+        }
+      },
+    };
+  },
+});
+
+interface IdentifierCounts {
+  general: number;
+  return: number;
+}
+
+type IdentifierArea = keyof IdentifierCounts;
+
+/**
+ * Count uses of type parameters in inferred return types.
+ * We need to resolve and analyze the inferred return type of a function
+ * to see whether it contains additional references to the type parameters.
+ * For classes, we need to do this for all their methods.
+ */
+function countTypeParameterUsage(
+  checker: ts.TypeChecker,
+  node: NodeWithTypeParameters,
+): Map<ts.Identifier, IdentifierCounts> {
+  const counts = new Map<ts.Identifier, IdentifierCounts>();
+
+  if (ts.isClassLike(node)) {
+    for (const typeParameter of node.typeParameters) {
+      collectTypeParameterUsageCounts(checker, typeParameter, counts);
+    }
+    for (const member of node.members) {
+      collectTypeParameterUsageCounts(checker, member, counts);
+    }
+  } else {
+    collectTypeParameterUsageCounts(checker, node, counts);
+  }
+
+  return counts;
+}
+
+/**
+ * Fills in a counter of the number of times each type parameter appears in
+ * the given type by recursively descending through type references.
+ * This is essentially a limited subset of the scope manager, but for types.
+ */
+function collectTypeParameterUsageCounts(
+  checker: ts.TypeChecker,
+  node: ts.Node,
+  identifierCounts: Map<ts.Identifier, IdentifierCounts>,
+): void {
+  const visitedSymbolLists = new Set<ts.Symbol[]>();
+  const type = checker.getTypeAtLocation(node);
+  const typeCounts = new Map<ts.Type, number>();
+  const visitedConstraints = new Set<ts.TypeNode>();
+  let functionLikeType = false;
+  let visitedDefault = false;
+
+  if (
+    ts.isCallSignatureDeclaration(node) ||
+    ts.isConstructorDeclaration(node)
+  ) {
+    functionLikeType = true;
+    visitSignature(checker.getSignatureFromDeclaration(node), 'general');
+  }
+
+  if (!functionLikeType) {
+    visitType(type, 'general', false);
+  }
+
+  function visitType(
+    type: ts.Type | undefined,
+    area: IdentifierArea,
+    asRepeatedType: boolean,
```

Comment:
Blurgh yeah I had a hard time naming it. Roughly, it's whether the type parameter is in a place that could _repeat_ its usage. In the file that's as an index signature or type argument. Any place like that essentially could repeat the type.

Reply:
Blurgh yeah I had a hard time naming it. Roughly, it's whether the type parameter is in a place that could _repeat_ its usage. In the file that's as an index signature or type argument. Any place like that essentially could repeat the type.

Reply:
I've read the code and your comment a few times and I still really don't understand what's going on here. Can you give an example of a tuple type where `collectTypeParameterUsage` would return the wrong number without this? i.e. if you changed:

```ts
    // Tuple types like `[K, V]`
    // Generic type references like `Map<K, V>`
    else if (tsutils.isTupleType(type) || tsutils.isTypeReference(type)) {
      for (const typeArgument of type.typeArguments ?? []) {
        visitType(typeArgument, true);
      }
    }
```

to have `asRepeatedType` or `false` instead of `true`.